internal/auth/oidc/service_start_auth.go

```diff
@@ -0,0 +1,143 @@
+// client can use to retrieve the results of the user's OIDC authentication
+// attempt. The tokenUrl contains a token_request_id, which is encrypted.
+//
+// If the auth method is in an InactiveState, then an error is returned.
+//
+// Options supported:
+//
+// WithRoundTripPayload(string) provides an option for a client roundtrip
+// payload.  This payload will be added to the final redirect as a query
+// parameter.
+func StartAuth(ctx context.Context, oidcRepoFn OidcRepoFactory, apiAddr string, authMethodId string, opt ...Option) (authUrl *url.URL, tokenUrl *url.URL, e error) {
+	const op = "oidc.StartAuth"
+	if apiAddr == "" {
+		return nil, nil, errors.New(errors.InvalidParameter, op, "missing api address")
+	}
+	if authMethodId == "" {
+		return nil, nil, errors.New(errors.InvalidParameter, op, "missing auth method id")
+	}
+	if oidcRepoFn == nil {
+		return nil, nil, errors.New(errors.InvalidParameter, op, "missing oidc repo function")
+	}
+	r, err := oidcRepoFn()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	am, err := r.lookupAuthMethod(ctx, authMethodId)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	if am == nil {
+		return nil, nil, errors.New(errors.RecordNotFound, op, fmt.Sprintf("auth method %s not found", authMethodId))
+	}
+	if am.OperationalState == string(InactiveState) {
+		return nil, nil, errors.New(errors.AuthMethodInactive, op, "not allowed to start authentication attempt")
+	}
+
+	// get the provider from the cache (if possible)
+	provider, err := providerCache().get(ctx, am)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	callbackRedirect := fmt.Sprintf(CallbackEndpoint, apiAddr, authMethodId)
+
+	opts := getOpts(opt...)
+	finalRedirect := fmt.Sprintf(FinalRedirectEndpoint, apiAddr)
+	if opts.withRoundtripPayload != "" {
+		finalRedirect = fmt.Sprintf("%s?%s", finalRedirect, opts.withRoundtripPayload)
+	}
+	now := time.Now()
+	createTime, err := ptypes.TimestampProto(now.Truncate(time.Second))
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "create timestamp failed", errors.WithWrap(err))
+	}
+	exp, err := ptypes.TimestampProto(now.Add(AttemptExpiration).Truncate(time.Second))
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "exp timestamp failed", errors.WithWrap(err))
+	}
+	tokenRequestId, err := authtoken.NewAuthTokenId()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	nonce, err := oidc.NewID()
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to generate nonce", errors.WithWrap(err))
+	}
+	hash, err := provider.ConfigHash()
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to get provider config hash", errors.WithWrap(err))
+	}
+	st := &request.State{
+		TokenRequestId:     tokenRequestId,
+		CreateTime:         &timestamp.Timestamp{Timestamp: createTime},
+		ExpirationTime:     &timestamp.Timestamp{Timestamp: exp},
```

Comment:
I think the appropriate time to decide when to expire is at the time the callback or polling request are made and not at the time the start authenticate returns.  Consider only including the create time and leaving the calculation of the expiration time for the endpoints servicing these requests.

Reply:
I disagree.  The TTL will not change in-flight, since it's immutable based on this point in the authentication process.  Of course the callback and token request endpoints are free apply some sort of leeway if they deem it appropriate to the exp, but that doesn't change the time of expiration.

Reply:
BTW, CreateTime is purely informational and not required by any downstream actor, so you could argue for not including it in the payload, but I see it sort of like the `iat` claim.  It's just a bit of helpful info for downstream participants.